### PR TITLE
Styling Components Research

### PR DESCRIPTION
**Pull request by JakeDawkins**
## tldr;

I refactored
`imports/components/icons/Accounts.js`
`imports/components/accountType/index.js`
`imports/blocks/discover/Item.js` (incomplete)

In this PR, I only show comparisons of our current solution (Aphrodite) and the proposed solution (styled-components).

Aphrodite creates a single `<style>` block at the top of the page for styles. At runtime, the `css()` function generates a classname for each component. It does not have support for react-native, which is an issue. It does not support global CSS or all of the CSS spec. It does not support theming, and overriding isn't very straight-forward.

Styled-Components supports react-native (which is a great plus). It injects styles similar to how aphrodite does, but it supports all of the CSS spec as well as global CSS styles. It would take a bit of refactoring to convert to, though.

**Edit: This is possible, actually. Haven't done it yet.**

In the PR, I changed discover cards to use this (mostly) instead of aphrodite, and some things, like passing a new class to ImageLoader for making the images fill the space don't work. We would need to rethink how we pass that kind of customizability in. With aphrodite, we were doing: 

```
  bgClasses = () => (
    [
      "grid__item",
      "two-fifths",
      "hard",
      "soft-half-left",
      "background--cover",
      css(Styles["height-100"]),
    ]
```

and I changed it to:

```
  bgClasses = `
    grid__item
    two-fifths
    hard
    soft-half-left
    background--cover
  `;
```

screen shot 2016-11-22 at 4 23 42 pm

## Aphrodite:

> CSS is only injected into the <head> of the page when these classes are used. The injection happens in Aphrodite's css() function

* [why only a single `style` component](https://github.com/Khan/aphrodite/issues/45)

* no support for react native

* could abstract styles to a separate file by importing the `styles` object

* [Allows extension of classes](https://github.com/Khan/aphrodite#user-content-advanced-extensions), although they may break because of specificity

## styled-components:

* [Video explaining styled-components](https://youtu.be/19gqsBc_Cx0?t=29m00s). The beginning of the video compares other styling solutions.

* [Does support SSR](https://github.com/styled-components/styled-components/issues/124) as do all the other options.

* Supports react-native styling (with actual CSS) :party:

* Supports all valid CSS (including @font-face and @keyframe) 

* Uses [Glamor](https://github.com/threepointone/glamor) for style injection.

* Enforces style declarations outside of JSX, keeping render methods clean.

* [Has Flow type support](https://github.com/styled-components/styled-components/blob/master/docs/flow-support.md).

* A bit verbose, and involves creating a new component and caling that component.

## Conclusion

styled-components could be beneficial to our long-term development. As we eventually move Holzmann to react-native, we will need to use some solution to style our components, and being able to use the same styling language for both native and web components would make our work quite a bit quicker and less painful.

Theming might be extremely useful down the road, as we allow and assist other churches to use holzmann (or even when we use it for Fuse). It would allow us to define things like theme colors, borders, etc in one place making the change quick and easy.

**To figure out**

I'm not sure _yet_ how using styled-components will affect testing of components. I imagine it would at most require us exporting the component before being wrapped (which we already do with some components).

I'm also not certain what kind of problems we will run into with our existing CSS codebase and specificity? 

Some cases of aphrodite use will not be possible to exactly replicate. We'll have to rethink how we do it.

### CODE REVIEW

**Comment by jbaxleyiii** [comment]
**[Storybook Hub](https://hub.getstorybook.io/)** is building storybooks for every commit in this PR.

**Latest Storybook: https://ff01860f-ff8b-4536-bb77-26b1f41a6f2f.sbook.io/**
<sub>( Visit [here](https://hub.getstorybook.io/app/github.com/NewSpring/Holtzman/branch/try-styled-components/) for more info )</sub>

> _This comment is automatically added by Storybook Hub_

**Comment by mxstbr** [comment]
If you have any specific questions or concerns about `styled-components`, please let me know and I'll happily answer all of them!

*I got a [jsbin](https://jsbin.com/gesaniseto/edit?js,output) that can convert simple styles-as-objects to CSS, (e.g. `{ color: 'blue' }` to `color: blue;`) which could be helpful for converting the styles from `aphrodite` to `styled-components`! Please note that it's not very smart, but it does save quite a bit of time.*

**Comment by jbaxleyiii** [comment]
@mxstbr thanks so much for the link and just stopping by to say 👋! We are big fans of you and your work at NewSpring Web! We are really excited about styled-components especially as we are building an application platform for churches to be able to customize to their brand / needs. Styled components is a must to do that well!

**Comment by mxstbr** [comment]
That's great to hear, thank you! 😊

Theming is one of the main usecases we wanted to solve, so definitely let me know how it goes. Good luck with the project!
